File: sample/flutterApp/lib/src/screens/field_states_screen.dart
// Copyright © 2026 Shiva Thapa (@shivathapaa). All rights reserved.
// Use of this source code is governed by the MPL 2.0 license that can be
// found in the LICENSE file.

import 'package:flutter/material.dart';
import 'package:nepali_date_picker_kmp/nepali_date_picker_kmp.dart';

import '../support/demo_section.dart';
import '../support/sample_defaults.dart';

/// Error, disabled, read-only and localized field states, and every wire
/// pattern formatting and parsing on its own.
class FieldStatesScreen extends StatefulWidget {
  const FieldStatesScreen({super.key});

  @override
  State<FieldStatesScreen> createState() => _FieldStatesScreenState();
}

class _FieldStatesScreenState extends State<FieldStatesScreen> {
  static const _patterns = <(DatePattern, String)>[
    (DatePattern.yyyySlashMmSlashDd, 'YYYY/MM/DD'),
    (DatePattern.yyyyDashMmDashDd, 'YYYY-MM-DD'),
    (DatePattern.ddSlashMmSlashYyyy, 'DD/MM/YYYY'),
    (DatePattern.ddDashMmDashYyyy, 'DD-MM-YYYY'),
  ];

  SimpleDate? _today;
  bool _touched = false;
  bool _deliveryEmpty = true;
  String _formatted = '…';
  String _devanagari = '…';
  String _parsedBack = '…';
  String _misParsed = '…';
  String _rangeStart = 'none';
  String _rangeEnd = 'none';
  bool _endBeforeStart = false;

  bool get _showError => _touched && _deliveryEmpty;

  @override
  void initState() {
    super.initState();
    _load();
  }

  Future<void> _load() async {
    final today = (await NepaliDateConverter.todayBs()).toSimpleDate();
    const seed = SimpleDate(2081, 6, 12);
    final formatted = await NepaliDateFormatter.format(seed,
        pattern: DatePattern.yyyySlashMmSlashDd);
    final devanagari = await NepaliDateFormatter.format(seed,
        pattern: DatePattern.ddDashMmDashYyyy, script: DigitScript.devanagari);
    final parsedBack = await NepaliDateFormatter.parse(formatted,
        pattern: DatePattern.yyyySlashMmSlashDd);
    final misParsed = await NepaliDateFormatter.parse('2081/06/12',
        pattern: DatePattern.ddSlashMmSlashYyyy);
    if (!mounted) return;
    setState(() {
      _today = today;
      _formatted = formatted;
      _devanagari = devanagari;
      _parsedBack = parsedBack == null ? 'null' : dateText(parsedBack);
      _misParsed = misParsed == null ? 'null' : dateText(misParsed);
    });
  }

  @override
  Widget build(BuildContext context) {
    final today = _today;
    return DemoScreen(
      title: 'Field states',
      children: [
        DemoSection(
          title: 'Error and supporting text',
          subtitle: 'Only dates from today onwards parse; an incomplete '
              'entry flips the field into its error state.',
          child: Column(children: [
            if (today != null)
              NepaliDateField(
                label: 'Delivery date',
                selectableDates:
                    NepaliSelectableDates.after(today, includeDate: true),
                supportingText: _showError
                    ? 'Enter a complete date from today onwards'
                    : 'Format: YYYY/MM/DD',
                isError: _showError,
                onValueChanged: (date) => setState(() {
                  _touched = true;
                  _deliveryEmpty = date == null;
                }),
              ),
          ]),
        ),
        DemoSection(
          title: 'Every input pattern',
          subtitle: 'The four wire shapes as field masks, then the formatter '
              'on its own, Devanagari included.',
          child: Column(children: [
            for (final (pattern, label) in _patterns)
              Padding(
                padding: const EdgeInsets.only(bottom: 8),
                child: NepaliDateField(
                  label: label,
                  dateFormat: pattern,
                  onValueChanged: (_) {},
                ),
              ),
            LabeledValue('format 2081-06-12', _formatted),
            LabeledValue('devanagari, day first', _devanagari),
            LabeledValue('parsed back', _parsedBack),
            LabeledValue('read with the wrong pattern', _misParsed),
          ]),
        ),
        DemoSection(
          title: 'Disabled',
          subtitle: 'A locked value the user can see but not touch.',
          child: const NepaliDateField(
            enabled: false,
            initialValue: SimpleDate(2081, 2, 8),
            label: 'Locked date',
            supportingText: 'Unlocked once the form above is complete',
            onValueChanged: _ignore,
          ),
        ),
        DemoSection(
          title: 'Read only',
          subtitle: 'Focusable and copyable, never editable.',
          child: NepaliDateField(
            readOnly: true,
            initialValue: today,
            label: 'Booked for',
            onValueChanged: (_) {},
          ),
        ),
        DemoSection(
          title: 'The filled style and its own words',
          subtitle: 'Nepali labels, placeholder and dialog buttons on the '
              'filled variant.',
          child: NepaliDateField(
            outlined: false,
            locale: SampleDefaults.nepali,
            label: 'मिति',
            placeholder: 'वर्ष/महिना/गते',
            confirmButtonText: 'ठिक छ',
            dismissButtonText: 'रद्द',
            onValueChanged: (_) {},
          ),
        ),
        DemoSection(
          title: 'A range pair in every state',
          subtitle: 'The end flags itself when it falls before the start; '
              'both sides accept Devanagari digits.',
          child: Column(children: [
            NepaliDateRangeField(
              outlined: false,
              startLabel: 'Leave from',
              endLabel: 'Leave until',
              supportingText: _endBeforeStart
                  ? 'The end cannot fall before the start'
                  : 'Both sides accept Devanagari digits',
              isEndError: _endBeforeStart,
              onRangeChanged: (start, end) => setState(() {
                _rangeStart = start == null ? 'none' : dateText(start);
                _rangeEnd = end == null ? 'none' : dateText(end);
                _endBeforeStart =
                    start != null && end != null && end.compareTo(start) < 0;
              }),
            ),
            LabeledValue('Start', _rangeStart),
            LabeledValue('End', _rangeEnd),
          ]),
        ),
      ],
    );
  }
}

void _ignore(SimpleDate? _) {}

File: sample/iosSwiftApp/iosSwiftApp/FieldStatesScreen.swift
//
//  FieldStatesScreen.swift
//  Nepali Date Picker
//
//  What a typed field can be told about itself: the error it is in, the text under it, whether it
//  accepts input at all, and the words on the buttons of the calendar behind it.
//

import SwiftUI
import nepali_date_picker

struct FieldStatesScreen: View {
    @State private var validated: SimpleDate?
    @State private var touched = false
    @State private var patterned: [SimpleDate?] = Array(repeating: nil, count: 4)
    @State private var rangeStart: SimpleDate?
    @State private var rangeEnd: SimpleDate?

    private let converter = NepaliDateConverter.shared

    /// Every input pattern the library takes, in the order the enum declares them.
    private let patterns: [(NepaliDateFormatter.Pattern, String)] = [
        (.yyyySlashMmSlashDd, "YYYY/MM/DD"),
        (.yyyyDashMmDashDd, "YYYY-MM-DD"),
        (.ddSlashMmSlashYyyy, "DD/MM/YYYY"),
        (.ddDashMmDashYyyy, "DD-MM-YYYY")
    ]

    /// A rule the validation demo judges against, so the error state has a reason to appear.
    private var futureOnly: NepaliSelectableDates {
        converter.AfterDateSelectable(simpleDate: converter.todayNepaliSimpleDate, includeDate: true)
    }

    var body: some View {
        ScrollView {
            VStack(spacing: 16) {
                validationSection
                patternsSection
                disabledSection
                readOnlySection
                dialogWordsSection
                rangeStatesSection
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 16)
        }
        .navigationTitle("Field states")
        .navigationBarTitleDisplayMode(.inline)
    }

    private var validationSection: some View {
        DemoSection(
            title: "Error and supporting text",
            subtitle: "A field reports null until what is typed is a complete date that passes the rule it was given. Drive isError from that, and put the reason in supportingText. This one only accepts today or later."
        ) {
            AutoSized(measurementHeight: 160) { report in
                NepaliDateFieldView(
                    onHeightChange: report,
                    initialValue: nil,
                    selectableDates: futureOnly,
                    label: "Delivery date",
                    supportingText: showError
                        ? "Enter a complete date from today onwards"
                        : "Format: YYYY/MM/DD",
                    isError: showError
                ) { value in
                    touched = true
                    validated = value
                }
            }
            .id(showError)
            LabeledValue(label: "Accepted", value: validated?.text ?? "none")
        }
    }

    private var showError: Bool { touched && validated == nil }

    private var patternsSection: some View {
        DemoSection(
            title: "Every input pattern",
            subtitle: "The pattern fixes the field order and the separator for both the mask and the text written back. NepaliDateFormatter formats and parses the same four without a field at all."
        ) {
            ForEach(Array(patterns.enumerated()), id: \.offset) { index, entry in
                AutoSized(measurementHeight: 140) { report in
                    NepaliDateFieldView(
                        onHeightChange: report,
                        initialValue: nil,
                        dateFormat: entry.0,
                        label: entry.1
                    ) { patterned[index] = $0 }
                }
                LabeledValue(label: entry.1, value: patterned[index]?.text ?? "incomplete")
            }
            Divider().padding(.horizontal, 12)
            ForEach(Array(patterns.enumerated()), id: \.offset) { _, entry in
                LabeledValue(
                    label: "format \(entry.1)",
                    value: NepaliDateFormatter.shared.format(
                        date: converter.todayNepaliSimpleDate,
                        pattern: entry.0,
                        script: .latin
                    )
                )
            }
            LabeledValue(
                label: "Devanagari digits",
                value: NepaliDateFormatter.shared.format(
                    date: converter.todayNepaliSimpleDate,
                    pattern: .yyyySlashMmSlashDd,
                    script: .devanagari
                )
            )
            LabeledValue(
                label: "parse 2081/06/12",
                value: NepaliDateFormatter.shared
                    .parse(input: "2081/06/12", pattern: .yyyySlashMmSlashDd)?.text ?? "no match"
            )
            LabeledValue(
                label: "parse the same as DD/MM",
                value: NepaliDateFormatter.shared
                    .parse(input: "2081/06/12", pattern: .ddSlashMmSlashYyyy)?.text ?? "no match"
            )
        }
    }

    private var disabledSection: some View {
        DemoSection(
            title: "Disabled",
            subtitle: "enabled = false greys the field and stops both typing and the calendar button, for a form waiting on something else to be filled in first."
        ) {
            AutoSized(measurementHeight: 140) { report in
                NepaliDateFieldView(
                    onHeightChange: report,
                    initialValue: SimpleDate(year: 2081, month: 2, dayOfMonth: 8),
                    label: "Locked date",
                    supportingText: "Unlocked once the form above is complete",
                    enabled: false
                ) { _ in }
            }
        }
    }

    private var readOnlySection: some View {
        DemoSection(
            title: "Read only",
            subtitle: "readOnly keeps the field looking ordinary and focusable but refuses edits, which is the shape a confirmation screen wants: the value is legible and copyable, and it cannot be changed by accident."
        ) {
            AutoSized(measurementHeight: 140) { report in
                NepaliDateFieldView(
                    onHeightChange: report,
                    initialValue: converter.todayNepaliSimpleDate,
                    label: "Booked for",
                    readOnly: true
                ) { _ in }
            }
        }
    }

    private var dialogWordsSection: some View {
        DemoSection(
            title: "The filled style and its own words",
            subtitle: "outlined = false gives the filled Material style, whose trailing icon opens a calendar. The two button labels and the corner radius are the field's own, so a localized app never shows an English OK."
        ) {
            AutoSized(measurementHeight: 140) { report in
                NepaliDateFieldView(
                    onHeightChange: report,
                    initialValue: nil,
                    locale: SampleDefaults.nepali,
                    outlined: false,
                    label: "मिति",
                    placeholder: "वर्ष/महिना/गते",
                    confirmButtonText: "ठिक छ",
                    dismissButtonText: "रद्द",
                    cornerRadius: 20
                ) { _ in }
            }
        }
    }

    private var rangeStatesSection: some View {
        DemoSection(
            title: "A range pair in every state",
            subtitle: "The two sides carry their own labels and their own error flags, because a range is wrong on one end at a time. This one marks the end red until it is on or after the start."
        ) {
            AutoSized(measurementHeight: 180) { report in
                NepaliDateRangeFieldView(
                    onHeightChange: report,
                    initialStartValue: nil,
                    initialEndValue: nil,
                    outlined: false,
                    startLabel: "Leave from",
                    endLabel: "Leave until",
                    supportingText: endBeforeStart
                        ? "The end cannot fall before the start"
                        : "Both sides accept Devanagari digits",
                    isStartError: false,
                    isEndError: endBeforeStart,
                    cornerRadius: 20
                ) { start, end in
                    rangeStart = start
                    rangeEnd = end
                }
            }
            .id(endBeforeStart)
            LabeledValue(label: "Start", value: rangeStart?.text ?? "incomplete")
            LabeledValue(label: "End", value: rangeEnd?.text ?? "incomplete")
        }
    }

    /// True once both ends are complete and the end is the earlier of the two.
    private var endBeforeStart: Bool {
        guard let start = rangeStart, let end = rangeEnd else { return false }
        return converter.getNepaliDaysInBetween(startDate: start, endDate: end) < 0
    }
}

#Preview {
    NavigationStack { FieldStatesScreen() }
}
